File: Stack/SinglyLinkedList.cpp
#include <iostream>
using namespace std;

class node{
    public:
        int data;
        node * next;
};

class SinglyLinkedList{
    private:
    node *head, *tail;
    int length;
    public:
    SinglyLinkedList(){
        head = NULL;
        tail = NULL;
        length = 0;
    }

    void append(int data){
        node *newNode = new node;
        (*newNode).data = data;
        (*newNode).next = NULL;

        if(head==NULL){
            head = newNode;
            tail = newNode;
        }else{
            (*tail).next = newNode;
            tail = newNode;
        }
        length ++;

    }

    void prepend(int data){
        node *newNode = new node;
        newNode->data = data;
        if(head==NULL){
            head = newNode;
            tail = newNode;
        }else{
            newNode->next = head;
            head = newNode;
        }
        length ++;
    }

    node * transverse_to_index(int index){
        node *curr = head;
        int count = 0;
        while(curr != NULL){
            if(count < index){
                curr = curr->next;
                count++;
            }else{
                break;
            }
        }
        return curr;
    }

    void eliminate(int index){
        if(head == NULL){
            throw "EmptyList";
        }else if (index < 0 || index >= length ) {
            throw "OutOfBound";
        }else if(index == 0 ){
            if(length == 1){
                head = NULL;
                tail = NULL;
            }else{
                node * curr = head;
                head = head->next;
                curr->next = NULL;
                curr = NULL;

            }
        }else{
            node *at= transverse_to_index(index);
            node *before = transverse_to_index(index - 1);
            if (index == length - 1){
                tail = before;
                before->next = NULL;
                at = NULL;
            }else{
                before->next = before->next->next;
                at->next=NULL;
                at=NULL;
            }
            
        }
        length--;
    }

    int lookup(int data){
        node * curr = head;
        int count = 0;
        while(curr!=NULL){
            if(data != curr->data){
                curr = curr->next;
                count++;
            }else{
                return count;
            }
        }
        return -1;
    }

    void display(){
        node *curr = head;
        cout<<"Length of "<< length <<endl;
        cout<<"head->";
        while(curr!=NULL){
            cout << curr->data << "->";
            curr = curr->next;
        }
        cout<<"NULL\n";
    }

};

int main(){
    SinglyLinkedList sll;
    sll.append(1);
    sll.append(2);
    sll.append(3);
    sll.append(4);
    sll.append(5);
    sll.append(6);
    sll.prepend(7);
    sll.prepend(8);
    sll.prepend(9);
    sll.prepend(10);
    sll.display();
    sll.eliminate(0);
    sll.eliminate(8);
    sll.eliminate(2);
    printf("Index %d\n",sll.lookup(9));
    sll.display();

    return 0;
}

File: Stack/SinglyLinkedList.py

class OutOfBoundLinkedList(Exception):
    pass

class SinglyLinkedListNode:
    def __init__(self,data = None,next= None):
        self.data = data
        self.next = next

class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def prepend(self,data):
        self.prepend_node(SinglyLinkedListNode(data))

    def append(self,data):
        self.append_node(SinglyLinkedListNode(data))

    def prepend_node(self,node):
        if self.head == None:
            self.head = node
            self.tail = node
        else:
            node.next = self.head
            self.head = node
        self.length += 1

    def append_node(self,node):
        if self.tail == None:
            self.tail = node
            self.head = node
        else:
            self.tail.next = node
            self.tail = node
        self.length += 1    

    def insert_node_by_index(self,index,node):
        if self.head == None: #empty 
            self.head = node
        elif index == 0: #empty 
            self.prepend(node)
        elif index == self.length: #at end, expect one item beyond max length
            self.append(node)
        elif index > self.length or index < 0: #at end, expect beyond max length 
            raise OutOfBoundLinkedList
        else:
            curr = self.head 
            counter = 0
            while curr != None:
                curr = curr.next
                counter += 1
                if counter == index:
                    node.next = curr.next
                    curr.next = node


    def insert_data_by_index(self,index,data):
        node = SinglyLinkedListNode(data,None)
        self.insert_node_by_index(index,node)


    def delete(self,index):
        #Conds for existence
        value = -1 
        if self.length  < 1: # empty 
            return value
        elif index > self.length - 1: # expect beyond max length 
            return value 
        else:
            #Algor
            curr = self.head 
            counter = 0
            if index == 0: #index = 0 problem
                if self.length == 1:
                    self.head = None
                else:
                    self.head = self.head.next
            else: #stand-at-front approach with one cursor
                while curr != None: 
                    curr = curr.next
                    counter += 1
                    if counter == index - 1:
                        value = curr.next.data
                        if counter == self.length - 2:
                            curr.next = None 
                        else:
                            curr.next = curr.next.next
        
        return value

    def lookup(self,data):
        index = 0
        if self.length <= 0: # empty problem
            return -1
        else:
            curr = self.head
            if curr.data == data:
                index = 0
            else:
                while curr != None:
                    curr = curr.next
                    index += 1
                    if curr.data == data:
                        return index 
        

    def display(self):
        curr = self.head
        print("Length of {}".format(self.length))
        print('Head->',end = '')
        if self.length == 1: #or self.head.next == None
            print(curr.data)
        else:
            while curr != None: 
                print (str(curr.data) + '->',end ='')
                curr= curr.next
        print('None')

    def reverse(self):
        pass
        

if __name__ == '__main__':
    sll = SinglyLinkedList()
    sll.append(7)
    sll.append(8)
    sll.append(9)
    sll.append(7)
    sll.prepend(1)
    sll.append(83)
    sll.append(9)
    sll.display()
    index = sll.lookup(83)

    sll.insert_data_by_index(6,11)
    sll.display()
    sll.delete(0)
    
    print(index)
    sll.display()
    

